### PR TITLE
Add MMLU tests

### LINKED ISSUE #15: Add mmlu testing code
Add the mmlu testing code in a sep file 

convo from: https://github.com/instructlab/eval/pull/10#discussion_r1651266896

### PR DESCRIPTION
**Pull request by nathan-weinberg**
Resolves #15

### CODE REVIEW

**Comment by nathan-weinberg** [comment]
Successfully ran for both MMLU:
```shell
(venv) [ec2-user@ip-10-0-14-244 tests (test)]$ python3.11 test_mmlu.py 
===> Executing test1...
2024-06-27:15:39:45,393 INFO     [evaluator.py:131] Setting random seed to 0 | Setting numpy seed to 1234 | Setting torch manual seed to 1234
2024-06-27:15:39:45,431 INFO     [huggingface.py:162] Using device 'cuda'
/home/ec2-user/nweinber/venv/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████| 3/3 [00:02<00:00,  1.03it/s]
You are using the default legacy behaviour of the <class 'transformers.models.llama.tokenization_llama_fast.LlamaTokenizerFast'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
2024-06-27:15:39:53,646 WARNING  [evaluator.py:222] Overwriting default num_fewshot of mmlu_astronomy from None to 2
2024-06-27:15:39:53,646 WARNING  [evaluator.py:222] Overwriting default num_fewshot of mmlu_anatomy from None to 2
2024-06-27:15:39:53,647 INFO     [task.py:395] Building contexts for mmlu_astronomy on rank 0...
100%|████████████████████████████████████████████████████████████████████████████████████████████████| 152/152 [00:00<00:00, 162.80it/s]
2024-06-27:15:39:54,588 INFO     [task.py:395] Building contexts for mmlu_anatomy on rank 0...
100%|████████████████████████████████████████████████████████████████████████████████████████████████| 135/135 [00:00<00:00, 163.34it/s]
2024-06-27:15:39:55,422 INFO     [evaluator.py:362] Running loglikelihood requests
Running loglikelihood requests: 100%|███████████████████████████████████████████████████████████████| 1148/1148 [00:23<00:00, 48.84it/s]
0.501827485380117
{'mmlu_anatomy': {'score': 0.4444444444444444, 'stderr': 0.04292596718256981}, 'mmlu_astronomy': {'score': 0.5592105263157895, 'stderr': 0.04040311062490436}}
```
And MMLUBranch
```shell
(venv) [ec2-user@ip-10-0-14-244 tests (test)]$ python3.11 test_mmlubranch.py 
===> Executing test1...
2024-06-27:15:50:18,495 INFO     [evaluator.py:131] Setting random seed to 0 | Setting numpy seed to 1234 | Setting torch manual seed to 1234
2024-06-27:15:50:18,534 INFO     [huggingface.py:162] Using device 'cuda'
/home/ec2-user/nweinber/venv/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████| 3/3 [00:02<00:00,  1.03it/s]
You are using the default legacy behaviour of the <class 'transformers.models.llama.tokenization_llama_fast.LlamaTokenizerFast'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
2024-06-27:15:50:21,826 WARNING  [task.py:322] [Task: softball] has_training_docs and has_validation_docs are False, using test_docs as fewshot_docs but this is not recommended.
2024-06-27:15:50:21,826 WARNING  [task.py:322] [Task: softball] has_training_docs and has_validation_docs are False, using test_docs as fewshot_docs but this is not recommended.
2024-06-27:15:50:21,892 WARNING  [task.py:322] [Task: tonsils] has_training_docs and has_validation_docs are False, using test_docs as fewshot_docs but this is not recommended.
2024-06-27:15:50:21,892 WARNING  [task.py:322] [Task: tonsils] has_training_docs and has_validation_docs are False, using test_docs as fewshot_docs but this is not recommended.
2024-06-27:15:50:21,897 WARNING  [evaluator.py:222] Overwriting default num_fewshot of tonsils from None to 2
2024-06-27:15:50:21,897 WARNING  [evaluator.py:222] Overwriting default num_fewshot of softball from None to 2
2024-06-27:15:50:21,897 INFO     [task.py:395] Building contexts for tonsils on rank 0...
100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 27/27 [00:00<00:00, 175.55it/s]
2024-06-27:15:50:22,055 INFO     [task.py:395] Building contexts for softball on rank 0...
100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 24/24 [00:00<00:00, 172.83it/s]
2024-06-27:15:50:22,197 INFO     [evaluator.py:362] Running loglikelihood requests
Running loglikelihood requests: 100%|█████████████████████████████████████████████████████████████████| 204/204 [00:03<00:00, 53.59it/s]
0.7450980392156863
{'mmlu_pr': {'score': 0.7450980392156863, 'stderr': 0.0619412614751624}}
```

File: tests/test_mmlubranch.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        overall_score, individual_scores = mmlu.run()
+        print(overall_score)
+        print(individual_scores)
+    except Exception as exc:
+        print(f"test1 failed: {exc}")
+        return False
+    return True
+
+
+if __name__ == "__main__":
```
Comment (JamesKunstle) [inline tests/test_mmlubranch.py:26]: Is having a test-runner like pytest on the roadmap at some point? Not pressuring, purely curious.

File: tests/testdata/sdg/softball_data.jsonl
Comment (JamesKunstle) [inline tests/testdata/sdg/softball_data.jsonl:1]: this is tonsil data not softball data ⚾

**Review by JamesKunstle — CHANGES_REQUESTED** [review]

File: tests/test_mmlubranch.py (hunk 1 above)
Comment (nathan-weinberg) [reply tests/test_mmlubranch.py:26]: Yes I'd love to make these proper unit tests at some point!

File: tests/testdata/sdg/softball_data.jsonl
Comment (nathan-weinberg) [reply tests/testdata/sdg/softball_data.jsonl:1]: Good catch! :gloves:

Comment (nathan-weinberg) [reply tests/testdata/sdg/softball_data.jsonl:1]: @alimaredia do you have that data? I'm not seeing it on the VM where I pulled these files from

Comment (nathan-weinberg) [reply tests/testdata/sdg/softball_data.jsonl:1]: I'm just going to remove these for now

File: tests/testdata/sdg/tonsil_task.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+dataset_kwargs:
+  data_files:
+    test: tonsil_data.jsonl
```
Comment (alimaredia) [inline tests/testdata/sdg/tonsil_task.yaml:3]: This is usually the absolute path to the `.jsonl` file for the task, but in our case a relative path will work as long as we can assume the dir the test is going to run out of. Take a look at https://github.com/instructlab/instructlab/blob/main/tests/testdata/mmlu_branch/tonsils_task.yaml#L3

In our case I think it's fine to assume running test_mmlubranch.py will be done out of the root of the eval repo, so the path here needs to be `test: tests/testdata/sdg/tonsil_data.jsonl`

**Review by alimaredia — CHANGES_REQUESTED** [review]

File: tests/test_mmlu.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# First Party
+from instructlab.eval.mmlu import MMLUEvaluator
+
+
+def test1():
```
Comment (danmcp) [inline tests/test_mmlu.py:5]: Can this be named something more descriptive.  Maybe test_minimal_mmlu?

File: tests/test_mmlubranch.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Standard
+import os
+
+# First Party
+from instructlab.eval.mmlu import MMLUBranchEvaluator
+
+
+def test1():
```
Comment (danmcp) [inline tests/test_mmlubranch.py:8]: maybe test_mmlu_branch

**Comment by danmcp** [comment]
Would be good to squash commits before we merge.  Happy to give my approval after.

**Review by danmcp — APPROVED** [review]

**Review by alinaryan — APPROVED** [review]

**Pull request merged**